Add DB "helper" script

jrm86 opened the pull request:
This PR addresses #10. It establishes a `DatabaseHelper` class which will be used by the app to interact with the database. The DB is established, and tables are created.

A few considerations:
- To complete insert, update, and delete operations, classes would have to be written for each of our db entities (`user`, `arc`, and `task`). This may be jumping a bit more ahead than we'd like, since the details of the rest of the app are still in planning/design at this point.
- Data types are limited in SQLite, and table attributes are dynamically typed, hence the many check constraints. Also, there is no `TIMESTAMP` equivalent.
- SQLite does not have a built-in way to use regex, so check constraints for something like an email address or a `TIMESTAMP` may have to be handled within app functions. Considering this is not a multi-tiered architecture, this shouldn't be a problem in theory, but it would be wise to keep in mind when creating the insert/update functions.

jrm86 committed:
add databaseHelper file

jrm86 committed:
add sqflite and pathprovider dependencies

jrm86 committed:
Add create tables

jrm86 committed:
Add constraints to DB

jrm86 committed:
Add blank files for DB objects

For write, update, delete operations, classes will be needed to interact with the DB

jrm86 committed:
add preamble and change data types

KevinKelly25 commented:
The PR is coming along great @jrm86, thank you. 

I am admittedly still learning a lot about dart/flutter/SQLite so take everything with a grain of salt. I found a couple things that may need to be changed.

- The UUID attributes should be `<` rather then `<=`

- It may be worthwhile to make a variable for all UUIDs for the max length such as `const int UUIDSize = 60`. This will be helpful for when we inevitably have to change UUID size since we currently have no implemented method of UUID generation and most are no 60 characters. Creating a constant variable for the other max lengths may be helpful as well since they are all set to 60 as well and could change.

jrm86 commented:
Thanks for the review, @KevinKelly25.  I'll look into your comments shortly.

I've created issue #18 to address the need for Insert, Update, and Delete operations. At the moment, we need other parts of the design to be complete before continuing. Once this PR is merged, it should be sufficient to close #10.  Closes #10

jrm86 commented:
Pardon the close/reopen. I blame my laptop's track-pad.

jrm86 committed:
add constants for attribute lengths

jrm86 commented:
@KevinKelly25, I've addressed your comments in the latest commit.

jrm86 commented:
Justin and I brought up a couple points in PR #11, which would also affect how the DB is implemented here. It's a rather quick fix once we make a decision in that thread.  Direct link to discussion [here](https://github.com/smacademic/project-cgkm/pull/11#issuecomment-474985692).

jgrabowski121 commented:
This PR is looking good upon my initial review @jrm86 . 
I noticed two things that may need to be fixed
-1. The table name for User was changed to ArcUser in the file `createServerTables.sql` so we should        
     aim to keep this consistant
-2. I was under the impression that the ID attributes @KevinKelly25 mentioned need to be a length of 
      exactly 60 as stated in our wiki for t[able and attribute summaries](https://github.com/smacademic/project-cgkm/wiki/Logical-Schema-and-Table-Summaries).

KevinKelly25 commented:
@jgrabowski121 is correct, I meant to say:

The `UUID` attributes should be `=` rather then `<=`

I am sorry for the mixup

jrm86 commented:
Good catch @jgrabowski121.  I should have caught it, too.  Commit on it's way shortly.

jrm86 committed:
change name of user table, adjust constraints

Changed table name from "user" to "arcuser".  Adjusted UUID constraint to == instead of <.  Added constant for email size

KevinKelly25 reviewed (approved):
Thank you for fixing the issue @jrm86. It looks good to merge now.

jgrabowski121 reviewed (approved):
Thank you @jrm86  for finalizing the changes. I will approve the PR it looks good to merge

MatthewChastain reviewed (approved):
In looking through the files I see no errors. Good implementation using SQLite's more limited functionality as compared to Postgres. Looks good!
